zulipterminal/ui_tools/boxes.py

```diff
@@ -95,53 +96,83 @@ def stream_box_view(self, caption: str='', title: str='') -> None:
-        except (IndexError, TypeError):
-            return None
+        for prefix, autocomplete_func in autocomplete_map.items():
+            if text.startswith(prefix):
+                self.typeahead_mode = True
+
+                typeaheads = autocomplete_func(text, prefix)
+                reduced_typeaheads = typeaheads[:num_suggestions]
+                truncated = len(reduced_typeaheads) != len(typeaheads)
+                try:
+                    typeahead = reduced_typeaheads[state]
+                except (IndexError, TypeError):
+                    typeahead = (None, None)  # type: ignore
+
+                self.view.set_typeahead_footer(reduced_typeaheads,
+                                               typeahead[1],
+                                               truncated)
+                return typeahead[0]
+
+        return text
 
     def autocomplete_mentions(self, text: str, prefix_string: str
-                              ) -> List[str]:
+                              ) -> List[Tuple[str, str]]:
         # Handles user mentions (@ mentions and silent mentions)
         # and group mentions.
-        group_typeahead = ['@*{}*'.format(group_name)
-                           for group_name in self.model.user_group_names
-                           if match_group(group_name, text[1:])]
+        groups = [group_name
+                  for group_name in self.model.user_group_names
+                  if match_group(group_name, text[1:])]
+        group_typeahead = format_string(groups, '@*{}*')
 
         users_list = self.view.users
-        user_typeahead = [prefix_string + '**{}**'.format(user['full_name'])
-                          for user in users_list
-                          if match_user(user, text[len(prefix_string):])]
+        users = [user['full_name']
+                 for user in users_list
+                 if match_user(user, text[len(prefix_string):])]
+        user_typeahead = format_string(users, prefix_string + '**{}**')
+
         combined_typeahead = group_typeahead + user_typeahead
 
         return combined_typeahead
 
-    def autocomplete_streams(self, text: str) -> List[str]:
+    def autocomplete_streams(self, text: str, prefix_string: str
+                             ) -> List[Tuple[str, str]]:
         streams_list = self.view.pinned_streams + self.view.unpinned_streams
-        stream_typeahead = [('#**{}**'.format(stream[0]), stream[0])
-                            for stream in streams_list]
-        return match_stream(stream_typeahead, text[1:],
-                            self.view.pinned_streams)
+        streams = [stream[0]
+                   for stream in streams_list]
+        stream_typeahead = format_string(streams, '#**{}**')
+
+        matched_typeaheads = match_stream(stream_typeahead, text[1:],
+                                          self.view.pinned_streams)
+        typeahead_data = list(map(lambda x: (x, x[3:-2]), matched_typeaheads))
```

Comment:
I still have to do this hacky stuff to generate the structure again.
It would be a lot more useful if `match_stream` only took a list of stream names and returned the matched list. Styling of the list could be done after that. But I guess, there would be problems in changing `match_streams` this due to the buttons?

Reply:
I think it would be reasonable to adapt `match_stream` return the text it is matching against too. That's a pretty minimal change, particularly if we later generalize `match_stream` to be useful for other matching, which seems likely?